sites/upsun/src/learn/overview/app-life-cycle.md

```diff
@@ -0,0 +1,105 @@
+---
+title: The application lifecycle
+weight: 3
+description: "Understand the {{% vendor/name %}} application lifecycle and learn how to use build, deploy, and runtime hooks to control app behavior"
+---
+<!-- vale off -->
+Hooks let you run custom commands at specific points in your application’s lifecycle: during [`build`, `deploy`](/learn/overview/build-deploy.html), and at `runtime`. They’re essential for setting up your app, managing graceful shutdowns, or preparing instances to handle traffic in autoscaling environments.
+
+Each web application on {{% vendor/name %}} can define web commands inside its `.upsun/config.yaml` file:
+
+```yaml
+applications:
+  myapp:
+    type: 'python:3.13'
+    web:
+      commands:
+        pre_start: "./scripts/setup.sh"
+        start: "uwsgi --ini conf/server.ini"
+        post_start: "./scripts/warm-up.sh"
+```
+## Web commands
+
+
+| Name          | Type   | Required | Blocks Traffic | Description |
+|---------------|--------|----------|------------------|-------------|
+| **pre_start** | string | No       | No               | Runs just before `start`. Useful for short per-instance setup actions, such as moving cache files or setting permissions. |
+| **start**     | string | Yes      | Yes (until running) | The main command that launches your app. The instance cannot serve traffic until `start` is successfully running. If it terminates, {{% vendor/name %}} restarts it immediately. |
+| **post_start**| string | No       | Yes (until completed) | Runs after the `start` command but *before* the container is added to the router. Instances will not receive traffic until this script completes successfully, making it ideal for warm-up tasks that must finish before the app begins handling requests. |
+
+For more information about web commands, visit the [Single-runtime Image page](/create-apps/app-reference/single-runtime-image.html#web-commands).
+
+{{< note theme="info" title="Blocking vs. non-blocking hooks" >}}
+
+A hook is considered *blocking* if the instance cannot receive traffic until that hook finishes successfully.  
+
+- `start` blocks traffic until the application is running.  
+- `post_start` blocks traffic until all warm-up tasks complete.  
+
+Non-blocking hooks, such as `pre_start`, run before traffic concerns and do not affect when an instance begins receiving requests.
```

Comment:
technically, the `pre_start` runs before the `start`, so if `start` blocks, then `pre_start` also does :/